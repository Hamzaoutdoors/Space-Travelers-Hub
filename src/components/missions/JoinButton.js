import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { Button } from 'react-bootstrap';
import { joinMission, leaveMission } from '../../redux/missions/missions';

const JoinButton = ({ status, id }) => {
  const dispatch = useDispatch();
  const fields = {
    'Not a member': {
      variant: 'outline-secondary',
      message: 'Join Mission',
      click: () => { dispatch(joinMission(id)); },
    },
    'Active member': {
      variant: 'outline-danger',
      message: 'Leave Mission',
      click: () => { dispatch(leaveMission(id)); },
    },
  };

  return (
    <Button
      className="text-nowrap mx-3"
      variant={fields[status].variant}
      onClick={fields[status].click}
    >
      {fields[status].message}
    </Button>
  );
};

JoinButton.propTypes = {
  id: PropTypes.string.isRequired,
  status: PropTypes.oneOf(['Not a member', 'Active member']).isRequired,
};

export default JoinButton;
