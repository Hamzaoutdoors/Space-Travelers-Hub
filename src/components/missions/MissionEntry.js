import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';

const joinMessage = (status) => {
  if (status === 'Not a member') {
    return (
      <Button
        className="text-nowrap mx-3"
        variant="outline-secondary"
      >
        Join Mission
      </Button>
    );
  }

  return (
    <Button
      className="text-nowrap mx-3"
      variant="outline-danger"
    >
      Leave Mission
    </Button>
  );
};

const statusBadge = (status) => {
  const bg = status === 'Not a member' ? 'secondary' : 'primary';
  const classes = status === 'Not a member' ? 'mx-2 text-uppercase' : 'mx-2';
  return <Badge className={classes} bg={bg}>{status}</Badge>;
};

const MissionEntry = ({
  id, name, description, status,
}) => (
  <tr key={id}>
    <td>{name}</td>
    <td>{description}</td>
    <td className="align-middle">{statusBadge(status)}</td>
    <td className="align-middle">{joinMessage(status)}</td>
  </tr>
);

MissionEntry.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
};

export default MissionEntry;
