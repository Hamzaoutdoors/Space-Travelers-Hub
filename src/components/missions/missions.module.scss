.buttonStyle {
  min-width: 135px;
}
