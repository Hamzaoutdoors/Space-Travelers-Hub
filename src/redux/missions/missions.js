import fetchMissions from '../../utils/missions-api';

const FETCH_MISSIONS = 'spaceX/missions/FETCH_MISSIONS';
const ADD_MISSIONS = 'spaceX/missions/ADD_MISSIONS';
const API_SUCCESS = 'spaceX/missions/API_SUCCESS';
const API_FAILURE = 'spaceX/missions/API_FAILURE';

const initialMissions = [];

export const fetchMissionsAction = () => (dispatch) => {
  dispatch({ type: FETCH_MISSIONS });
  return fetchMissions().then(
    (missions) => {
      dispatch({ type: API_SUCCESS });
      dispatch({ type: ADD_MISSIONS, missions });
    },
    (error) => {
      dispatch({ type: API_FAILURE, error });
    },
  );
};

const missionsReducer = (state = initialMissions, action) => {
  switch (action.type) {
    case ADD_MISSIONS:
      return [...state, ...action.missions];
    default:
      return state;
  }
};

export default missionsReducer;
