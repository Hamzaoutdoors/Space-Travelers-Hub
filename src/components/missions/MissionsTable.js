import Table from 'react-bootstrap/Table';
import { shallowEqual, useSelector } from 'react-redux';
import MissionEntry from './MissionEntry';

const MissionsTable = () => {
  const missions = useSelector((state) => state.missions, shallowEqual);

  return (
    <Table striped bordered hover size="sm" responsive>
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody>
        {missions.map(({
          id, name, description, status,
        }) => (
          <MissionEntry
            key={id}
            id={id}
            name={name}
            description={description}
            status={status}
          />
        ))}
      </tbody>
    </Table>
  );
};

export default MissionsTable;
