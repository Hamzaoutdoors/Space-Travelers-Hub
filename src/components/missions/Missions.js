import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Container from 'react-bootstrap/Container';
import MissionsTable from './MissionsTable';
import { fetchMissionsAction } from '../../redux/missions/missions';

let initialized = false;

const Missions = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (initialized) return;
    dispatch(fetchMissionsAction());
    initialized = true;
  }, []);

  return (
    <Container className="missions">
      <MissionsTable />
    </Container>
  );
};

export default Missions;
