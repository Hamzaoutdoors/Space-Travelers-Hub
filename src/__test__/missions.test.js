import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render, fireEvent, screen, cleanup } from '@testing-library/react';
import { within } from '@testing-library/dom'
import MissionsTable from '../components/missions/MissionsTable';
import store from '../redux/configureStore';
import { fetchMissionsAction } from '../redux/missions/missions';

jest.mock('../utils/missions-api');

describe('Missions tests', () => {
  it('Missions page matches snapshot', async () => {
    await store.dispatch(fetchMissionsAction());
    const missionsPage = renderer
      .create(<Provider store={store}><MissionsTable /></Provider>)
      .toJSON();
    expect(missionsPage).toMatchSnapshot();
  });

  describe('Interaction tests', () => {
    beforeEach(() => {
      render(<Provider store={store}><MissionsTable /></Provider>)
    });
    afterEach(cleanup);

    it('Changes button after clicking button', () => {
      const thaicomRow = screen.getByText('Thaicom').closest('tr');

      expect(within(thaicomRow).queryByText('Leave Mission')).not
        .toBeInTheDocument();
      const joinButton = within(thaicomRow).getByText('Join Mission');
      fireEvent.click(joinButton);
      expect(within(thaicomRow).queryByText('Leave Mission'))
        .toBeInTheDocument();
      fireEvent.click(joinButton);
      expect(within(thaicomRow).queryByText('Leave Mission')).not
        .toBeInTheDocument();
    });

    it('Changes badge after clicking button', () => {
      const telstarRow = screen.getByText('Telstar').closest('tr');

      expect(within(telstarRow).queryByText('Active member')).not
        .toBeInTheDocument();
      const joinButton = within(telstarRow).getByText('Join Mission');
      fireEvent.click(joinButton);
      expect(within(telstarRow).queryByText('Active member'))
        .toBeInTheDocument();
      fireEvent.click(joinButton);
      expect(within(telstarRow).queryByText('Active member')).not
        .toBeInTheDocument();
    });
  });
});
