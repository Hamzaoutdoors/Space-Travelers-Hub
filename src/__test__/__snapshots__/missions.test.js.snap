// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions tests Missions page matches snapshot 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-sm table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <td />
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Thaicom
        </td>
        <td>
          Thaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited, which is the company that owns and operates the Thaicom satellite fleet and other telecommunication businesses in Thailand and throughout the Asia-Pacific region. The satellite projects were named Thaicom by the King of Thailand, His Majesty the King Bhumibol Adulyadej, as a symbol of the linkage between Thailand and modern communications technology.
        </td>
        <td
          className="align-middle"
        >
          <p
            className="badge mx-2 text-uppercase bg-secondary my-0"
            style={
              Object {
                "transform": "rotateX(360deg) translateZ(0)",
              }
            }
          >
            Not a member
          </p>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="text-nowrap mx-3 buttonStyle btn btn-outline-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Telstar
        </td>
        <td>
          Telstar 19V (Telstar 19 Vantage) is a communication satellite in the Telstar series of the Canadian satellite communications company Telesat. It was built by Space Systems Loral (MAXAR) and is based on the SSL-1300 bus. As of 26 July 2018, Telstar 19V is the heaviest commercial communications satellite ever launched, weighing at 7,076 kg (15,600 lbs) and surpassing the previous record, set by TerreStar-1 (6,910 kg/15230lbs), launched by Ariane 5ECA on 1 July 2009.
        </td>
        <td
          className="align-middle"
        >
          <p
            className="badge mx-2 text-uppercase bg-secondary my-0"
            style={
              Object {
                "transform": "rotateX(360deg) translateZ(0)",
              }
            }
          >
            Not a member
          </p>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="text-nowrap mx-3 buttonStyle btn btn-outline-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Iridium NEXT
        </td>
        <td>
          In 2017, Iridium began launching Iridium NEXT, a second-generation worldwide network of telecommunications satellites, consisting of 66 active satellites, with another nine in-orbit spares and six on-ground spares. These satellites will incorporate features such as data transmission that were not emphasized in the original design. The constellation will provide L-band data speeds of up to 128 kbit/s to mobile terminals, up to 1.5 Mbit/s to Iridium Pilot marine terminals, and high-speed Ka-band service of up to 8 Mbit/s to fixed/transportable terminals. The next-generation terminals and service are expected to be commercially available by the end of 2018. However, Iridium's proposed use of its next-generation satellites has raised concerns the service will harmfully interfere with GPS devices. The satellites will incorporate a secondary payload for Aireon, a space-qualified ADS-B data receiver. This is for use by air traffic control and, via FlightAware, for use by airlines. A tertiary payload on 58 satellites is a marine AIS ship-tracker receiver, for Canadian company exactEarth Ltd. Iridium can also be used to provide a data link to other satellites in space, enabling command and control of other space assets regardless of the position of ground stations and gateways.
        </td>
        <td
          className="align-middle"
        >
          <p
            className="badge mx-2 text-uppercase bg-secondary my-0"
            style={
              Object {
                "transform": "rotateX(360deg) translateZ(0)",
              }
            }
          >
            Not a member
          </p>
        </td>
        <td
          className="align-middle"
        >
          <button
            className="text-nowrap mx-3 buttonStyle btn btn-outline-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
